Model/Service/Sync/Upsert/Category/AsyncBulkConsumer.php

```diff
@@ -0,0 +1,109 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @author Nosto Solutions Ltd <contact@nosto.com>
+ * @copyright 2020 Nosto Solutions Ltd
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Tagging\Model\Service\Sync\Upsert\Category;
+
+use Magento\Framework\EntityManager\EntityManager;
+use Magento\Framework\Json\Helper\Data as JsonHelper;
+use Magento\Store\Model\App\Emulation;
+use Nosto\Exception\MemoryOutOfBoundsException;
+use Nosto\NostoException;
+use Nosto\Tagging\Helper\Scope as NostoScopeHelper;
+use Nosto\Tagging\Logger\Logger;
+use Nosto\Tagging\Model\ResourceModel\Magento\Product\CollectionFactory;
+use Nosto\Tagging\Model\Service\Sync\AbstractBulkConsumer;
+use Nosto\Tagging\Model\Service\Sync\Upsert\SyncService;
+
+/**
+ * Asynchronous Bulk Consumer
+ *
+ * Class AsyncBulkConsumer
+ */
+class AsyncBulkConsumer extends AbstractBulkConsumer
+{
+    /** @var SyncService */
+    private SyncService $syncService;
+
+    /** @var NostoScopeHelper */
+    private NostoScopeHelper $nostoScopeHelper;
+
+    /** @var CollectionFactory */
+    private CollectionFactory $collectionFactory;
+
+    /**
+     * AsyncBulkConsumer constructor.
+     * @param SyncService $syncService
+     * @param NostoScopeHelper $nostoScopeHelper
+     * @param CollectionFactory $collectionFactory
+     * @param JsonHelper $jsonHelper
+     * @param EntityManager $entityManager
+     * @param Emulation $storeEmulation
+     * @param Logger $logger
+     */
+    public function __construct(
+        SyncService $syncService,
+        NostoScopeHelper $nostoScopeHelper,
+        CollectionFactory $collectionFactory,
+        JsonHelper $jsonHelper,
+        EntityManager $entityManager,
+        Emulation $storeEmulation,
+        Logger $logger
+    ) {
+        $this->syncService = $syncService;
+        $this->nostoScopeHelper = $nostoScopeHelper;
+        $this->collectionFactory = $collectionFactory;
+        parent::__construct(
+            $logger,
+            $jsonHelper,
+            $entityManager,
+            $storeEmulation
+        );
+    }
+
+    /**
+     * @inheritDoc
+     * @throws MemoryOutOfBoundsException
+     * @throws NostoException
+     */
+    public function doOperation(array $productIds, string $storeId)
+    {
+        $store = $this->nostoScopeHelper->getStore($storeId);
+        $productCollection = $this->collectionFactory->create()
+            ->addIdsToFilter($productIds)
+            ->addStoreFilter($storeId);
+        $this->syncService->syncProducts($productCollection, $store);
```

Comment:
@ugljesaspx 
You're sending categories, not products. Please check this properly and which syncService is this, should be a syncService for categories, not products.